crates/reqwest/src/lib.rs

```diff
@@ -211,3 +211,307 @@ impl<S: CacheStorage> reqwest_middleware::Middleware for Cache<S> {
+    use http_cache_stream::storage::DefaultCacheStorage;
+    use reqwest::Response;
+    use reqwest::StatusCode;
+    use reqwest::header;
+    use reqwest_middleware::ClientWithMiddleware;
+    use reqwest_middleware::Middleware;
+    use tempfile::tempdir;
+
+    use super::*;
+
+    struct MockMiddlewareState {
+        responses: Vec<Option<Response>>,
+        current: usize,
+    }
+
+    struct MockMiddleware(Mutex<MockMiddlewareState>);
+
+    impl MockMiddleware {
+        fn new<R>(responses: impl IntoIterator<Item = R>) -> Self
+        where
+            R: Into<Response>,
+        {
+            Self(Mutex::new(MockMiddlewareState {
+                responses: responses.into_iter().map(|r| Some(r.into())).collect(),
+                current: 0,
+            }))
+        }
+    }
+
+    impl Middleware for MockMiddleware {
+        fn handle<'a, 'b, 'c, 'd>(
+            &'a self,
+            _: Request,
+            _: &'b mut Extensions,
+            _: Next<'c>,
+        ) -> BoxFuture<'d, reqwest_middleware::Result<Response>>
+        where
+            'a: 'd,
+            'b: 'd,
+            'c: 'd,
+            Self: 'd,
+        {
+            async {
+                let mut state = self.0.lock().unwrap();
+
+                let current = state.current;
+                state.current += 1;
+
+                Ok(state
+                    .responses
+                    .get_mut(current)
+                    .expect("unexpected client request: not enough responses defined")
+                    .take()
+                    .unwrap())
+            }
+            .boxed()
+        }
+    }
+
+    #[tokio::test]
+    async fn no_store() {
+        const BODY: &str = "hello world!";
+        const DIGEST: &str = "3aa61c409fd7717c9d9c639202af2fae470c0ef669be7ba2caea5779cb534e9d";
+
+        let dir = tempdir().unwrap();
+        let cache = Arc::new(Cache::new(DefaultCacheStorage::new(dir.path())));
+        let mock = Arc::new(MockMiddleware::new([
+            http::Response::builder()
+                .header(header::CACHE_CONTROL, "no-store")
+                .body(BODY)
+                .unwrap(),
+            http::Response::builder()
+                .header(header::CACHE_CONTROL, "no-store")
+                .body(BODY)
+                .unwrap(),
+        ]));
+        let client = ClientWithMiddleware::new(
+            Default::default(),
+            vec![cache.clone() as Arc<dyn Middleware>, mock.clone()],
+        );
+
+        // Response should not be served from the cache or stored
+        let response = client.get("http://test.local/").send().await.unwrap();
+        assert_eq!(
+            response.headers().get(header::CACHE_CONTROL).unwrap(),
+            "no-store"
+        );
+        assert_eq!(response.headers().get(X_CACHE_LOOKUP).unwrap(), "MISS");
+        assert_eq!(response.headers().get(X_CACHE).unwrap(), "MISS");
+        assert!(response.headers().get(X_CACHE_DIGEST).is_none());
```

Comment:
Definitely beside the point of this PR, but we should be using `cache-status` instead of `x-cache` headers unless needed for compatibility with another system; see https://www.rfc-editor.org/rfc/rfc9211.html